pg_search/src/postgres/customscan/hook.rs

```diff
@@ -211,3 +216,235 @@ pub extern "C-unwind" fn paradedb_upper_paths_callback<CS>(
+    pg_sys::planner_hook = Some(paradedb_planner_hook);
+}
+
+/// Planner hook that replaces WindowFunc nodes before PostgreSQL processes them
+/// Only replaces window functions if the query will be handled by our custom scans
+#[pg_guard]
+unsafe extern "C-unwind" fn paradedb_planner_hook(
+    parse: *mut pg_sys::Query,
+    query_string: *const ::core::ffi::c_char,
+    cursor_options: ::core::ffi::c_int,
+    bound_params: pg_sys::ParamListInfo,
+) -> *mut pg_sys::PlannedStmt {
+    // Check if this is a SELECT query with window functions that we can handle
+    if !parse.is_null() && (*parse).commandType == pg_sys::CmdType::CMD_SELECT {
+        // Note: it's important to check for window functions first, then search operator,
+        // otherwise we'd call query_has_search_operator() during the DROP EXTENSION, which
+        // would cause a panic.
+        if query_has_window_functions(parse) && query_has_search_operator(parse) {
+            // Extract and replace window functions recursively (including subqueries)
+            replace_windowfuncs_recursively(parse);
+        }
+    }
+
+    // Call the previous planner hook or standard planner
+    static mut PREV_PLANNER_HOOK: pg_sys::planner_hook_type = None;
+    if let Some(prev_hook) = PREV_PLANNER_HOOK {
+        prev_hook(parse, query_string, cursor_options, bound_params)
+    } else {
+        pg_sys::standard_planner(parse, query_string, cursor_options, bound_params)
+    }
+}
+
+/// Check if the target list contains any window functions
+unsafe fn has_window_functions(target_list: *mut pg_sys::List) -> bool {
+    let tlist = PgList::<pg_sys::TargetEntry>::from_pg(target_list);
+    for te in tlist.iter_ptr() {
+        if nodecast!(WindowFunc, T_WindowFunc, (*te).expr).is_some() {
+            return true;
+        }
+    }
+    false
+}
+
+/// Check if the query (or any subquery) contains window functions
+unsafe fn query_has_window_functions(parse: *mut pg_sys::Query) -> bool {
+    if parse.is_null() {
+        return false;
+    }
+
+    // Check the current query's target list
+    if !(*parse).targetList.is_null() && has_window_functions((*parse).targetList) {
+        return true;
+    }
+
+    // Check subqueries in RTEs (only if SUBQUERY_SUPPORT is enabled)
+    if window_agg::window_functions::SUBQUERY_SUPPORT && !(*parse).rtable.is_null() {
+        let rtable = PgList::<pg_sys::RangeTblEntry>::from_pg((*parse).rtable);
+        for (idx, rte) in rtable.iter_ptr().enumerate() {
+            if (*rte).rtekind == pg_sys::RTEKind::RTE_SUBQUERY
+                && !(*rte).subquery.is_null()
+                && query_has_window_functions((*rte).subquery)
+            {
+                return true;
+            }
+        }
+    }
+
+    false
+}
+
+/// Check if the query contains the @@@ search operator
+/// This indicates that our custom scans will likely handle this query
+unsafe fn query_has_search_operator(parse: *mut pg_sys::Query) -> bool {
```

Comment:
I think that it we can use `create_upper_paths_hook` for this instead, then we don't need to do the function replacement, and can continue to use the code that we have in each custom scan for locating our operator, etc.